util/tracing: update tracing to have configurable sampling rate and batch sizes

jayshrivastava opened the pull request:
**Commit 1 - util/tracing: update tracing to have configurable batch size**

Release note (general change): Adds the cluster setting
trace.zipkin.collector.batch.size which allows the size
of batches sent from the collector to be configured.

**Commit 2 - util/tracing: update tracing to have configurable sampling rate**

Resolves: https://github.com/cockroachdb/cockroach/issues/45891

Release note (general change): Adds the cluster setting
trace.sampling.rate which can be used to set a sampling rate for
traces in the /debug page, Lightstep traces, and
Zipkin traces. For example, a value of 0.25 corresponds 25% of
all traces being recorded and collected.

**Commit 3 - util/tracing: refactored cluster settings listener**

Benchmarks: https://docs.google.com/document/d/10dB_GsqBu9WsAJ-0M1j43-fQk3SDVBbX7H8e3AV6ji4/edit?usp=sharing

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54346)
<!-- Reviewable:end -->

jayshrivastava committed:
util/tracing: update tracing to have configurable sampling rate

Resolves: #45891

Release note (general change): Adds the cluster setting
trace.sampling.rate which can be used to set a sampling rate for
traces in the /debug page, Lightstep traces, and
Zipkin traces. For example, a value of 0.25 corresponds 25% of
all traces being recorded and collected.

jayshrivastava committed:
util/tracing: update tracing to have configurable batch size

Release note (general change): Adds the cluster setting
trace.zipkin.collector.batch.size which allows the size
of batches sent from the collector to be configured.

jayshrivastava commented on pkg/server/settingsworker_test.go:
```diff
@@ -293,3 +305,186 @@ func TestSettingsShowAll(t *testing.T) {
+	setClusterTraceSamplingRate := func(rate float64) {
+		execStatement(fmt.Sprintf(`SET CLUSTER SETTING trace.zipkin.sampling.rate = %f`, rate))
+	}
+	execStatement(`SET CLUSTER SETTING trace.zipkin.collector.batch.size = 100`)
```
I actually did not end up setting the batch size to 0 to make the traces sent to the collector more synchronous. When writing 1000 rows in parallel, the collector simply drops traces if the buffer (which uses the same setting as the batch size) is full. Thus, the batch size needs to be large so traces don't get dropped and we can count them.

jayshrivastava commented on pkg/server/settingsworker_test.go:
```diff
@@ -375,3 +379,112 @@ func TestSettingsZipkinBatchSize(t *testing.T) {
+	// Set up tracing collector and variable to count the number of traces
+	// collected from INSERT INTO statements
+	nTraces := int64(0)
+	collectorHandlerFcn := func(rw http.ResponseWriter, req *http.Request) {
```
All this function does is increment `nTraces` only if we see an `INSERT INTO` statement.

jayshrivastava commented on pkg/server/settingsworker_test.go:
```diff
@@ -375,3 +379,112 @@ func TestSettingsZipkinBatchSize(t *testing.T) {
+	write10000Rows()
+	testutils.SucceedsSoon(t, func() error {
+		currentTraces := atomic.LoadInt64(&nTraces)
+		if currentTraces < 450 && currentTraces > 550 {
```
This has a bit of a race condition which may result in a false positive. It is possible that `currentTraces` may exceed 550 after we pass this test due to `nTraces` being incremented asynchronously. In testing, this basically never happened, which makes me feel safer about this.

jayshrivastava commented on pkg/server/settingsworker_test.go:
```diff
@@ -375,3 +379,111 @@ func TestSettingsZipkinBatchSize(t *testing.T) {
+	setClusterTraceSamplingRate := func(rate float64) {
+		execStatement(fmt.Sprintf(`SET CLUSTER SETTING trace.zipkin.sampling.rate = %f`, rate))
+	}
+	execStatement(fmt.Sprintf(`SET CLUSTER SETTING trace.zipkin.collector = '%s'`, traceCollectorServerURL.Host))
```
The test uses the default batch size of 100. Setting it to 0 actually caused problems because spans would get dropped when many were being created in parallel (apparently, the collector uses a buffer which has the same length as the batch size we set).

Although the sampling rate limiting feature did not use configurable batch sizes, I figured that the batch size setting does not hurt, and we might as well keep it as it might be useful in the future.

jayshrivastava commented on pkg/server/settingsworker_test.go:
```diff
@@ -293,3 +304,164 @@ func TestSettingsShowAll(t *testing.T) {
+	write1000Rows()
+	testutils.SucceedsSoon(t, func() error {
+		currentTraces := atomic.LoadInt64(&nTraces)
+		if currentTraces < 950 && currentTraces > 1050 {
```
Locally, I was able to get exactly 1000. In CI, this failed a few times. When I got less than 1000, I could see in the logs that the collector buffer would get filled and spans would get dropped. Sometimes, I would actually get over 1000, which I assume is due to retrying. 

If we really wanted exactly 1000 here, I could use map to track & count the unique keys being inserted. However, I don't think it is necessary because the main purpose of the test is to see if we can reduce the number of traces by setting a lower sampling rate. Checking to see that if we get roughly 1000 with a rate of 1.0 and then roughly 500 with a rate of 0.5 should be enough to test this.

ajwerner commented:
I have some bad news that is going to make this project dramatically more involved. What we have here is a great first step, however, it's probably not enough in the general case. The problem with this approach as is is that it will lead to cockroach tracing 100% of requests but only sending a small percentage to the trace collector. What we really want is to only trace the requests which will ultimately be sampled. That way we won't incur the overhead of tracing them all. Given that this PR is an incremental improvement and is strictly better than what we have now, we can move forward with it but I want to push you towards digging deeper into how we can determine whether a trace will end up being sampled, and, thus make it so that sampling is reasonable to run in production.

An end goal that would be valuable would be to be able to run TPC-C and with tracing on at a low sampling rate and to not observe a major decrease in performance.

jayshrivastava commented on pkg/sql/txn_state.go:
```diff
@@ -193,10 +193,6 @@ func (ts *txnState) resetForNewSQLTxn(
 
-	if !tracing.IsRecordable(sp) {
-		log.Fatalf(connCtx, "non-recordable transaction span of type: %T", sp)
-	}
```
It is possible that the span created might be a noop (non-recordable) now because of random sampling. We do not need to check this anymore.

jayshrivastava commented on pkg/util/tracing/tracer.go:
```diff
@@ -358,9 +410,14 @@ func (t *Tracer) AlwaysTrace() bool {
-		return &t.noopSpan
+
+	if !t.alwaysTrace() {
+		if recordable == NonRecordableSpan {
```
This if retains the same functionality as before

ajwerner reviewed (commented):
> Reviewed 4 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54346)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/server/settingsworker_test.go, line 405 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MHIY1eU_vSDDE0t92vs-r2-405:-MI0AtMFCh20ZUMRizbP:b3ufq2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0266d12d6c0500c18ae0ce8c9e8146de96a114cf/pkg/server/settingsworker_test.go#L405)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

All this function does is decode the body and is increment `nTraces` if we see an `INSERT INTO` statement. We also check that the state of the statement is `Open` to prevent double counting.
</blockquote></details>

can you say more about this double counting? when do we get more than one trace in different spans for the same statement? My abstract  fear with no clear justification, on some level, is that we're initializing tracing in what would be a child span after we've decided not to trace a root.
___
*[pkg/server/settingsworker_test.go, line 450 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MHIY1eU_vSDDE0t92vs-r5-450:-MI0BD-ABs2OFj-dWBAP:b-av6p0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/19719acaeea2498114cfac024653fa09e4c89060/pkg/server/settingsworker_test.go#L450)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

Locally, I was able to get exactly 1000. In CI, this failed a few times. When I got less than 1000, I could see in the logs that the collector buffer would get filled and spans would get dropped. Sometimes, I would actually get over 1000, which I assume is due to retrying. 

If we really wanted exactly 1000 here, I could use map to track & count the unique keys being inserted. However, I don't think it is necessary because the main purpose of the test is to see if we can reduce the number of traces by setting a lower sampling rate. Checking to see that if we get roughly 1000 with a rate of 1.0 and then roughly 500 with a rate of 0.5 should be enough to test this.
</blockquote></details>

yeah I think we need to make sure this isn't flakey but the property that it confirms can be loose.
___
*[pkg/server/settingsworker_test.go, line 310 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MI0AdZ9BHmaiOaBUDLy:-MI0AdZ9BHmaiOaBUDLz:bz071sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f98a452499504cc1afa17da6c44a9685d77c78/pkg/server/settingsworker_test.go#L310)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 
> 	// Set up collector handler function and batch size reporting
> ```

super nit: period at the end of sentences here and below
___
*[pkg/util/tracing/tracer.go, line 76 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MI0BMan2NxhdEzHceSj:-MI0BMan2NxhdEzHceSk:bk64o6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f98a452499504cc1afa17da6c44a9685d77c78/pkg/util/tracing/tracer.go#L76)):*
> ```Go
> 	"trace.zipkin.collector.batch.size",
> 	"if set, sets the maximum batch size after which traces will be collected",
> 	100,
> ```

was this the default in the zipkin collector? can you comment on where this comes from?
___
*[pkg/util/tracing/tracer.go, line 126 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MI0Bjgk69oBd0prLr3p:-MI0Bjgk69oBd0prLr3q:b-ej28fa) ([raw file](https://github.com/cockroachdb/cockroach/blob/23c8b5b4031d1d8dbb980797c0acaa28091b3ffe/pkg/util/tracing/tracer.go#L126)):*
> ```Go
> 
> 	// The fraction (between 0.0 and 1.0) of traces to record and collect stored
> 	// as an unsafe pointer for atomics. Set by changing the cluster settings.
> ```

note that it's a pointer to `*float64`
___
*[pkg/util/tracing/tracer.go, line 401 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MI0BweCEYzHoReXR-Zj:-MI0BweCEYzHoReXR-Zk:b-h4oz4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/23c8b5b4031d1d8dbb980797c0acaa28091b3ffe/pkg/util/tracing/tracer.go#L401)):*
> ```Go
> 		return false
> 	}
> 	return rand.Float64() < sr
> ```

how about:

```
return sr == 1 || (sr > 0 && rand.Float64() < sr)
```


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54346)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/server/settingsworker_test.go, line 405 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MHIY1eU_vSDDE0t92vs-r2-405:-MI0Ji9H1HYBN5Xz2c6O:bi7u2oe) ([raw file](https://github.com/cockroachdb/cockroach/blob/0266d12d6c0500c18ae0ce8c9e8146de96a114cf/pkg/server/settingsworker_test.go#L405)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can you say more about this double counting? when do we get more than one trace in different spans for the same statement? My abstract  fear with no clear justification, on some level, is that we're initializing tracing in what would be a child span after we've decided not to trace a root.
</blockquote></details>

Say I execute one "INSERT INTO ..." statement. In Jaeger, I would set the operation to "exec stmt" in the UI and hit "Find Traces". You can see what those traces are below. I believe each trace starts with a root span. 

![Screen Shot 2020-09-24 at 2.16.11 PM.png](https://files.reviewableusercontent.io/images/reviewable/18633281/xV0qzDUMOhnL66_RulHQp26U/Screen%20Shot%202020-09-24%20at%202.16.11%20PM.png)

I can show you what some of these look like:
![Screen Shot 2020-09-24 at 2.18.10 PM.png](https://files.reviewableusercontent.io/images/reviewable/18633281/EZdc0dHnwcTtBsar-oNa9r53/Screen%20Shot%202020-09-24%20at%202.18.10%20PM.png)![Screen Shot 2020-09-24 at 2.19.24 PM.png](https://files.reviewableusercontent.io/images/reviewable/18633281/mGCiMl0hab9myRvZx_uMAE34/Screen%20Shot%202020-09-24%20at%202.19.24%20PM.png)
![Screen Shot 2020-09-24 at 2.18.19 PM.png](https://files.reviewableusercontent.io/images/reviewable/18633281/2mSDvlwZ0_GHL2O0xZcFFtuO/Screen%20Shot%202020-09-24%20at%202.18.19%20PM.png)

You can see that the first 2 above are very similar, so I had to find a way to count each time a unique INSERT occurs. 
___
*[pkg/server/settingsworker_test.go, line 450 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MHIY1eU_vSDDE0t92vs-r5-450:-MI0M78N14SaCfTnu_5P:b-j0al0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/19719acaeea2498114cfac024653fa09e4c89060/pkg/server/settingsworker_test.go#L450)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

yeah I think we need to make sure this isn't flakey but the property that it confirms can be loose.
</blockquote></details>

I think this particular flakey behaviour was due to some older code. I haven't had any problems with it today. 
___
*[pkg/util/tracing/tracer.go, line 76 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MI0BMan2NxhdEzHceSj:-MI0J9_JDT2v_s1skI-p:bswycmk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f98a452499504cc1afa17da6c44a9685d77c78/pkg/util/tracing/tracer.go#L76)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

was this the default in the zipkin collector? can you comment on where this comes from?
</blockquote></details>

Yup the 100 comes from zipkintracer.defaultHTTPBatchSize. Unfortunately, it's not exported, so we cannot use it. 


<!-- Sent from Reviewable.io -->

jayshrivastava commented on pkg/util/tracing/tracer_span.go:
```diff
@@ -250,16 +249,6 @@ func (s *span) disableRecording() {
-func IsRecordable(os opentracing.Span) bool {
-	_, isCockroachSpan := os.(*span)
-	return isCockroachSpan
-}
```
Linter said this isn't used anywhere anymore

ajwerner reviewed (commented):
> Reviewed 1 of 5 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54346)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/rpc/context.go, line 103 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MI0_S_b0S0QAoaD9UjY:-MI0_S_b0S0QAoaD9UjZ:b-aq85o5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e3d4aca3e3516bac8548050607090406b272b20/pkg/rpc/context.go#L103)):*
> ```
> 	// Should we trace regardless of the client? This is useful for calls coming
> 	// through the HTTP->RPC gateway (i.e. the AdminUI), where client is never
> 	// tracing.
> ```
I'm curious about the impact here. My guess is that this is going to disable tracing of requests from the admin server which is not totally desirable. What about 
___
*[pkg/server/settingsworker_test.go, line 426 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MIOxlOh36eFBtO2iJw2:-MIOxlOh36eFBtO2iJw3:bjala6z) ([raw file](https://github.com/cockroachdb/cockroach/blob/60d17776431190d56a1a71cdae83be3b3f20487e/pkg/server/settingsworker_test.go#L426)):*
> ```Go
> 	execStatement(fmt.Sprintf(`SET CLUSTER SETTING trace.zipkin.collector = '%s'`, traceCollectorServerURL.Host))
> 
> 	// Function that writes 500 rows.
> ```

Let's move these counters to constants and tune them down under race so the test doesn't take so long. 
___
*[pkg/server/settingsworker_test.go, line 451 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MIOxtPN7Y-rLoF7CprJ:-MIOxtPN7Y-rLoF7CprK:b326cpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/60d17776431190d56a1a71cdae83be3b3f20487e/pkg/server/settingsworker_test.go#L451)):*
<details><summary><i alt="summary">Quoted 10 lines of code…</i></summary>

> ```
> 
> 	atomic.StoreInt64(&nTraces, 0)
> 	write1000Rows()
> 	testutils.SucceedsSoon(t, func() error {
> 		currentTraces := atomic.LoadInt64(&nTraces)
> 		if currentTraces < 475 && currentTraces > 525 {
> 			return fmt.Errorf(`could not trace all SQL 'INSERT INTO' statements. traced %d out of 500`, currentTraces)
> 		}
> 		return nil
> 	})
> ```
</details>


This feels like a good opportunity to use subtests with `t.Run` and to refactor this code to utilize a helper which takes the expectation bounds. 

consider:

```
numOps := 500
tolerance := 50
if util.RaceEnabled {
    numOps = 50
    tolerance = 20
}
type testCase struct {
    sampleRate float64
}
run := func(t *testing.T, tc testCase) {
    setClusterTraceSamplingRate(tc.sampleRate)
    atomic.StoreInt64(&nTraces, 0)
    writeRows(numOps)
    minAllowed := int64(float64(numOps)*tc.sampleRate - float64(tolerance)/2)
    maxAllowed := int64(float64(numOps)*tc.sampleRate + float64(tolerance)/2)
    testutils.SucceedsSoon(t, func() error {
         currentTraces := atomic.LoadInt64(&nTraces)
         if currentTraces < minAllowed || currentTraces > maxAllowed {
             return errors.Errorf("expected between %d and %d traces at a sample rate of %f of %d ops, got %d",
                  minAllowed, maxAllowed, tc.sampleRate, numOps, currentTraces)
         }
        return nil
    })
}
subtest := func(tc testCase) (string, func(t *testing.T)) {
    return fmt.Sprintf("sampleRate=%f", tc.sampleRate), func(t *testing.T) { run(t, tc) }
}
for _, tc := range []testcase{
    {1}, {.5}, {.1},
} {
    t.Run(subtest(tc))
}
```
___
*[pkg/server/settingsworker_test.go, line 461 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MIP-M_W5nOCDaq8LoVG:-MIP-M_W5nOCDaq8LoVH:b55py2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/60d17776431190d56a1a71cdae83be3b3f20487e/pkg/server/settingsworker_test.go#L461)):*
> ```Go
> 	testutils.SucceedsSoon(t, func() error {
> 		currentTraces := atomic.LoadInt64(&nTraces)
> 		if currentTraces < 225 && currentTraces > 275 {
> ```

should be `||`?
___
*[pkg/util/tracing/tracer.go, line 76 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MI0BMan2NxhdEzHceSj:-MI0_idB5QAEeKYJQKyL:bty9nw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f98a452499504cc1afa17da6c44a9685d77c78/pkg/util/tracing/tracer.go#L76)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

Yup the 100 comes from zipkintracer.defaultHTTPBatchSize. Unfortunately, it's not exported, so we cannot use it. 
</blockquote></details>

add some commentary please
___
*[pkg/util/tracing/tracer.go, line 149 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MI0_sQE6Ev0VGq-J1RL:-MI0_sQE6Ev0VGq-J1RM:bxrctd) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e3d4aca3e3516bac8548050607090406b272b20/pkg/util/tracing/tracer.go#L149)):*
<details><summary><i alt="summary">Quoted 19 lines of code…</i></summary>

> ```
> 	reconfigure := func() {
> 		if lsToken := lightstepToken.Get(sv); lsToken != "" {
> 			t.setShadowTracer(createLightStepTracer(lsToken))
> 		} else if zipkinAddr := zipkinCollector.Get(sv); zipkinAddr != "" {
> 			t.setShadowTracer(createZipkinTracer(
> 				zipkinAddr,
> 				zipkinCollectorBatchSize.Get(sv),
> 			))
> 		} else {
> 			t.setShadowTracer(nil, nil)
> 		}
> 		var nt int32
> 		if enableNetTrace.Get(sv) {
> 			nt = 1
> 		}
> 		atomic.StoreInt32(&t._useNetTrace, nt)
> 		sr := samplingRate.Get(sv)
> 		atomic.SwapPointer(&t._samplingRate, unsafe.Pointer(&sr))
> 	}
> ```
</details>

I'm a bit suspect about having just a single reconfigure function which we trigger for any of these variables changing. I guess it's fine. 


<!-- Sent from Reviewable.io -->

jayshrivastava committed:
util/tracing: refactored cluster settings listener

Release note: None

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54346)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/rpc/context.go, line 103 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MI0_S_b0S0QAoaD9UjY:-MIQXmto5litjPRuijyz:bax75yw) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e3d4aca3e3516bac8548050607090406b272b20/pkg/rpc/context.go#L103)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	// Should we trace regardless of the client? This is useful for calls coming
> 	// through the HTTP->RPC gateway (i.e. the AdminUI), where client is never
> 	// tracing.
> ```
I'm curious about the impact here. My guess is that this is going to disable tracing of requests from the admin server which is not totally desirable. What about 
</blockquote></details>

Fixed in latest commit. 
___
*[pkg/server/settingsworker_test.go, line 426 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MIOxlOh36eFBtO2iJw2:-MIQXvEj06uowsM7Il-b:bb27k1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/60d17776431190d56a1a71cdae83be3b3f20487e/pkg/server/settingsworker_test.go#L426)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Let's move these counters to constants and tune them down under race so the test doesn't take so long. 
</blockquote></details>

done in new commit
___
*[pkg/server/settingsworker_test.go, line 451 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MIOxtPN7Y-rLoF7CprJ:-MIQXyso8VXFVkcIoCfc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/60d17776431190d56a1a71cdae83be3b3f20487e/pkg/server/settingsworker_test.go#L451)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 
> 	atomic.StoreInt64(&nTraces, 0)
> 	write1000Rows()
> 	testutils.SucceedsSoon(t, func() error {
> 		currentTraces := atomic.LoadInt64(&nTraces)
> 		if currentTraces < 475 && currentTraces > 525 {
> 			return fmt.Errorf(`could not trace all SQL 'INSERT INTO' statements. traced %d out of 500`, currentTraces)
> 		}
> 		return nil
> 	})
> ```

This feels like a good opportunity to use subtests with `t.Run` and to refactor this code to utilize a helper which takes the expectation bounds. 

consider:

```
numOps := 500
tolerance := 50
if util.RaceEnabled {
    numOps = 50
    tolerance = 20
}
type testCase struct {
    sampleRate float64
}
run := func(t *testing.T, tc testCase) {
    setClusterTraceSamplingRate(tc.sampleRate)
    atomic.StoreInt64(&nTraces, 0)
    writeRows(numOps)
    minAllowed := int64(float64(numOps)*tc.sampleRate - float64(tolerance)/2)
    maxAllowed := int64(float64(numOps)*tc.sampleRate + float64(tolerance)/2)
    testutils.SucceedsSoon(t, func() error {
         currentTraces := atomic.LoadInt64(&nTraces)
         if currentTraces < minAllowed || currentTraces > maxAllowed {
             return errors.Errorf("expected between %d and %d traces at a sample rate of %f of %d ops, got %d",
                  minAllowed, maxAllowed, tc.sampleRate, numOps, currentTraces)
         }
        return nil
    })
}
subtest := func(tc testCase) (string, func(t *testing.T)) {
    return fmt.Sprintf("sampleRate=%f", tc.sampleRate), func(t *testing.T) { run(t, tc) }
}
for _, tc := range []testcase{
    {1}, {.5}, {.1},
} {
    t.Run(subtest(tc))
}
```
</blockquote></details>

Done.
___
*[pkg/server/settingsworker_test.go, line 461 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MIP-M_W5nOCDaq8LoVG:-MIQXy859tBk5-WJUgFY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/60d17776431190d56a1a71cdae83be3b3f20487e/pkg/server/settingsworker_test.go#L461)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

should be `||`?
</blockquote></details>

Done.
___
*[pkg/util/tracing/tracer.go, line 76 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MI0BMan2NxhdEzHceSj:-MIQYMNZ7OWfIfU8OFZi:bpire5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f98a452499504cc1afa17da6c44a9685d77c78/pkg/util/tracing/tracer.go#L76)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

add some commentary please
</blockquote></details>

added to the first commit
___
*[pkg/util/tracing/tracer.go, line 149 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MI0_sQE6Ev0VGq-J1RL:-MIQY14yASytvkqrmXgk:b-tihtbb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e3d4aca3e3516bac8548050607090406b272b20/pkg/util/tracing/tracer.go#L149)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	reconfigure := func() {
> 		if lsToken := lightstepToken.Get(sv); lsToken != "" {
> 			t.setShadowTracer(createLightStepTracer(lsToken))
> 		} else if zipkinAddr := zipkinCollector.Get(sv); zipkinAddr != "" {
> 			t.setShadowTracer(createZipkinTracer(
> 				zipkinAddr,
> 				zipkinCollectorBatchSize.Get(sv),
> 			))
> 		} else {
> 			t.setShadowTracer(nil, nil)
> 		}
> 		var nt int32
> 		if enableNetTrace.Get(sv) {
> 			nt = 1
> 		}
> 		atomic.StoreInt32(&t._useNetTrace, nt)
> 		sr := samplingRate.Get(sv)
> 		atomic.SwapPointer(&t._samplingRate, unsafe.Pointer(&sr))
> 	}
> ```
I'm a bit suspect about having just a single reconfigure function which we trigger for any of these variables changing. I guess it's fine. 
</blockquote></details>

Done in new commit


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> Reviewed 1 of 6 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54346)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/rpc/context.go, line 103 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MI0_S_b0S0QAoaD9UjY:-MIUGdnF1HRWRBiSurPK:b-x647z4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e3d4aca3e3516bac8548050607090406b272b20/pkg/rpc/context.go#L103)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

Fixed in latest commit. 
</blockquote></details>

move the comment above the return please. 
___
*[pkg/util/tracing/tracer.go, line 415 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MHC9u2AZwbwWCMHxN9E-r7-415:-MIUHylv7LPZSozALOiK:bemipj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/bce5400207c84f356843fa1372e2c83b8bc3b333/pkg/util/tracing/tracer.go#L415)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

This if retains the same functionality as before
</blockquote></details>

This smells wrong to me. Isn't the idea that if we're passing `NonRecordableSpan` then we want to sample the tracing if sampled tracing is enabled but if we're passing anything other than `NonRecordableSpan` then we definitely want to trace?
___
*[pkg/util/tracing/tracer.go, line 255 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MIUH6xP6_Z7bTOcobi2:-MIUH6xQ2TxvVohDf3eU:b-uskmlh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cc186d5d4bc1af00ff8259116892b77f330b461/pkg/util/tracing/tracer.go#L255)):*
> ```
> 	if !t.alwaysTrace() {
> 		if len(opts) == 0 && shadowTr == nil {
> ```
isn't this the same as `if !t.alwaysTrace() && len(opts) == 0 && shadowTr == nil {`?
___
*[pkg/util/tracing/tracer.go, line 303 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MIUHH3z5t0JyKm52iSJ:-MIUHH3z5t0JyKm52iSK:b1zflih) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cc186d5d4bc1af00ff8259116892b77f330b461/pkg/util/tracing/tracer.go#L303)):*
> ```Go
> 	if !hasParent {
> 		if !t.alwaysTrace() {
> 			if !t.shouldSample() {
> ```

isn't this the same as the conjunction?
___
*[pkg/util/tracing/tracer.go, line 318 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MIUHQCl0t16ijrOMVRz:-MIUHQCl0t16ijrOMVS-:b1zflih) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cc186d5d4bc1af00ff8259116892b77f330b461/pkg/util/tracing/tracer.go#L318)):*
> ```Go
> 	// part of a recording or snowball trace, avoid overhead and return a noop
> 	// span.
> 	if !t.alwaysTrace() {
> ```

isn't this the same as the conjunction?
___
*[pkg/util/tracing/tracer.go, line 400 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/54346#-MIUHn6M9iH2FVWmmk3G:-MIUHn6M9iH2FVWmmk3H:b-7etloe) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cc186d5d4bc1af00ff8259116892b77f330b461/pkg/util/tracing/tracer.go#L400)):*
> ```Go
> // alwaysTrace can be used to set conditions for when 100% tracing is desired.
> func (t *Tracer) alwaysTrace() bool {
> 	return t.useNetTrace() || t.forceRealSpans
> ```

I feel like this comment is wrong and name is wrong, right? does `useNetTrace()` imply 100% tracing?


<!-- Sent from Reviewable.io -->

jayshrivastava commented:
https://github.com/cockroachdb/cockroach/issues/45891
